### ISSUE TITLE
Split Tunneling + 'Block connections without VPN' as Firewall Risks

### ISSUE THREAD

**Issue by issuant** [mullvad/mullvadvpn-app#7197]
### I have checked if others have suggested this already

- [X] I have checked this issue tracker to see if others have reported similar issues.

### Feature description

It seems split tunneling an app while leaving 'Block connections without VPN' activated works as a makeshift firewall on Android. Is this safe? What privacy risks exist if this is used for any arbitrary app? Does this work on Windows too?

In https://github.com/mullvad/mullvadvpn-app/issues/7151#issuecomment-2476054121, it is said that apps cannot work outside the tunnel but reading the [security document](https://github.com/mullvad/mullvadvpn-app/blob/main/docs/split-tunneling.md), the full breadth of dangers for this 'hack' is not entirely clear.

Can a test, similar to the one that discovered the [DNS leaks](https://github.com/mullvad/mullvadvpn-app/blob/main/docs/known-issues.md), be done for this? If it does prove safe, this 'hack' could be a tolerable compromise for the many requests of an integrated firewall.

### Alternative solutions

n/a

### Type of feature

- [X] Better privacy/anonymity
- [ ] Better at circumventing censorship
- [ ] Easier to use
- [ ] Other

### Operating System

- [X] Android
- [ ] iOS
- [X] Windows
- [ ] macOS
- [ ] Linux

**Comment by MarkusPettersson98** [comment]
Hello! Could you clarify what you mean with "Does this work on Windows too?", I don't think I fully understand. On Windows, we already integrate with the firewall on Windows :blush:

**Comment by Serock3** [comment]
Hello! That's a clever solution to blocking apps from accessing the internet on Android!

I tested combining split tunneling with "Lockdown mode" on Windows, which is the closest equivalent setting. I can confirm that it does not
and no, we do not produce the same behavior. Apps that are split always have their traffic routed outside the tunnel.

**Comment by Serock3** [comment]
I feel like I need to clarify two things:
1. "Block connection without VPN" is an OS level settings on Android, which we have no control over. This is why it behaves differently from "Lockdown mode" on the desktop platforms.
2. Using the split tunneling feature to block apps is not a supported use case, do it at your own risk. It's unlikely to happen, but apps that are split from the tunnel could still "leak into the tunnel", and access the internet that way.

**Comment by Serock3** [comment]
> > I tested combining split tunneling with "Lockdown mode" on Windows, which is the closest equivalent setting. I can confirm that it does not produce the same behavior.
> 
> That is unfortunate, and a warning or notification should be added to the app informing users of this. The security document should also be updated. Does it still leak if the split-tunneled app ([#7072 (comment)](https://github.com/mullvad/mullvadvpn-app/issues/7072#issuecomment-2456037903)) is firewalled?

No, apps that are split from the tunnel will not leak if they are also blocked by you own firewall rules. Split tunneling simply means that we don't apply any firewall rules or routing entries for the selected programs, they will not be affected by the VPN app. This is very much the intended behavior. I have brought up the discussion with the app team about whether the interaction between split tunneling and lockdown mode can be communicated better. It isn't obvious for the end user that the Android behavior is different.

>   > Using the split tunneling feature to block apps is not a supported use case, do it at your own risk. It's unlikely to happen, but apps that are split from the tunnel could still "leak into the tunnel", and access the internet that way.
> 
> Robustly enumerating the risks, if any, is why I opened this. Brushing it aside as unsupported does not make much sense because even outside of the context of this 'hack', split tunneled apps on Android leaking when 'Block connections without VPN' is still active must be considered a significant bug, no less significant than the discovered DNS leaks.

We do not consider leaking traffic into the tunnel to be part of our threat model, since we are not a firewall. The app protects against leaking your IP address, which will be obscured as long as the traffic goes inside the tunnel.

**Comment by Serock3** [comment]
> Do you mean the apps that could not be firewalled in [#7072 (comment)](https://github.com/mullvad/mullvadvpn-app/issues/7072#issuecomment-2456037903) get firewalled if they are first split-tunneled?

Well, not exactly. Splitting apps from the tunnel will not help you avoid the difficulties of creating firewall rules that block their sub-processes. What I mean is that if you have a firewall rule that is able to block an app when the VPN is not active, it should function normally with split tunneling, or without it for that matter.

> While I would still consider leaks 'into the VPN' very much still a bug in this context, the privacy damage is a bit less severe.

If you have 'Block connections without VPN' enabled on Android you _cannot leak outside the tunnel_, even with split tunneling. So the answer is no, you don't have to worry about leaking when combining those settings. You can think of it as if the split tunneling "tricks" the apps to try to go outside the tunnel, which the OS firewall then detects and blocks. Hypothetically, apps can ignore this instruction and send their data inside the tunnel anyway. This is not ideal, but we have no way to prevent it.

On desktop platforms, apps that are split from the tunnel *will leak*. This is the entire purpose of the feature, of course. Any app that is split will go outside the tunnel at all times, regardless of 'Lockdown mode' or other settings. This discrepancy is a bit confusing, but there is little we can do to change get around system level settings on Android.

**Comment by MrChocolatine** [comment]
https://github.com/mullvad/mullvadvpn-app/issues/6775#issuecomment-2339058697

**Comment by MarkusPettersson98** [comment]
> I have seen those already. Neither provide any insight.

As per https://github.com/mullvad/mullvadvpn-app/issues/1747#issuecomment-628461792

> Implementing a firewall is out of scope for a VPN app, at least this VPN app and for the foreseeable future.

And https://github.com/mullvad/mullvadvpn-app/issues/6775#issuecomment-2410116849

> If you are asking about a configurable firewall within the app, then yes it is currently out of scope.
> 
> The app is using the default OS firewall on Windows, macOS and Linux for the killswitch and to prevent traffic leaking outside the tunnel.

@issuant  I think @Serock3 has provided you with great details on why this is the case on Android from a technical perspective. What more are you expecting, exactly?

**Comment by MarkusPettersson98** [comment]
It is out of scope because our product is not a firewall, it is a VPN app. We do integrate with the firewall on operating systems which provides their own firewalls to protect our users from network leaks. With that said, we don't have any intention of building our own firewall solution, since it would be a gigantious task for a company of our size. For context, we are currently 4 developers actively working on the system integration for all platforms but iOS.

As it stands, we have pretty good confidence that our service protects the anonymity of our users as intended because of audits and because we've been battle-tested for many years at this point. We have plenty of other areas that we strive to do better in, such as anti-censorship, were we'd rather dedicate our time at this moment.

I'll let you reply if have any feedback on what I just wrote, but I'll probably close this issue soon because we already have a standing feature request for adding a firewall. Thanks for your understanding :yellow_heart:

**Comment by MarkusPettersson98** [comment]
Yes, you do give up some level of control by using our app instead of vanilla Wireguard + rolling your own firewall rules. What you gain is a good set of vetted defaults for the majority of use cases with a no-compromise focus on privacy (e.g. preventing unwanted traffic leaks).

You are of course free to roll your own setup, but it is not something I recommend for anyone except advanced users that are aware of the risks of doing so. You can generate a Wireguard config via the account page on our website: https://mullvad.net/account/wireguard-config

As noted previously this setup won't work on Android, as you can only have one VPN active at a time , so you can't use the Wireguard together with any 'firewall' app which is just implemented as a VPN.

I'll go ahead and close this issue as we have other GitHub issues for tracking the interest of adding a firewall to the app. Please let me know if you disagree with this decision.
